benchmarks/functional_autograd_benchmark/audio_text_models.py

```diff
@@ -0,0 +1,397 @@
+        """
+
+        def __init__(self, d_model, dropout=0.1, max_len=5000):
+            super(PositionalEncoding, self).__init__()
+            self.dropout = nn.Dropout(p=dropout)
+
+            pe = torch.zeros(max_len, d_model)
+            position = torch.arange(0, max_len, dtype=torch.float).unsqueeze(1)
+            div_term = torch.exp(torch.arange(0, d_model, 2).float() * (-math.log(10000.0) / d_model))
+            pe[:, 0::2] = torch.sin(position * div_term)
+            pe[:, 1::2] = torch.cos(position * div_term)
+            pe = pe.unsqueeze(0).transpose(0, 1)
+            self.register_buffer('pe', pe)
+
+        def forward(self, x):
+            r"""Inputs of forward function
+            Args:
+                x: the sequence fed to the positional encoder model (required).
+            Shape:
+                x: [sequence length, batch size, embed dim]
+                output: [sequence length, batch size, embed dim]
+            Examples:
+                >>> output = pos_encoder(x)
+            """
+
+            x = x + self.pe[:x.size(0), :]
+            return self.dropout(x)
+
+    class TransformerModel(nn.Module):
+        """Container module with an encoder, a recurrent or transformer module, and a decoder."""
+
+        def __init__(self, ntoken, ninp, nhead, nhid, nlayers, dropout=0.5):
+            super(TransformerModel, self).__init__()
+            try:
+                from torch.nn import TransformerEncoder, TransformerEncoderLayer
+            except Exception:
+                raise ImportError('TransformerEncoder module does not exist in PyTorch 1.1 or lower.')
+            self.model_type = 'Transformer'
+            self.src_mask = None
+            self.pos_encoder = PositionalEncoding(ninp, dropout)
+            encoder_layers = TransformerEncoderLayer(ninp, nhead, nhid, dropout)
+            self.transformer_encoder = TransformerEncoder(encoder_layers, nlayers)
+            self.encoder = nn.Embedding(ntoken, ninp)
+            self.ninp = ninp
+            self.decoder = nn.Linear(ninp, ntoken)
+
+            self.init_weights()
+
+        def _generate_square_subsequent_mask(self, sz):
+            mask = (torch.triu(torch.ones(sz, sz)) == 1).transpose(0, 1)
+            mask = mask.float().masked_fill(mask == 0, float('-inf')).masked_fill(mask == 1, float(0.0))
+            return mask
+
+        def init_weights(self):
+            initrange = 0.1
+            nn.init.uniform_(self.encoder.weight, -initrange, initrange)
+            # Not sure how this works in the original code
+            # nn.init.zeros_(self.decoder)
+            nn.init.uniform_(self.decoder.weight, -initrange, initrange)
+
+        def forward(self, src, has_mask=True):
+            if has_mask:
+                device = src.device
+                if self.src_mask is None or self.src_mask.size(0) != len(src):
+                    mask = self._generate_square_subsequent_mask(len(src)).to(device)
```

Comment:
Why do we generate `mask` in each forward step? In a typical use, the mask is generated outside `forward` func.

Reply:
This was taken from the example here: https://github.com/pytorch/examples/blob/master/word_language_model/model.py#L108-L152

I can generate the mask outsite if you think it is more representative.

Reply:
It's not a good practice w.r.t. the performance. However, that example has to deal with RNN (without mask) and transformer models so I had to handle the mask within forward function. For most transformer case, users pass masks into the forward func of transformer.

Reply:
Ok! I'll update that.

Reply:
Since we always use the same batch size and the mask is cached. It will be created during the warmup forward and so should not cause any issue here!